trview.app/Lua/Elements/CameraSink/Lua_CameraSink.cpp

```diff
@@ -14,18 +17,97 @@ namespace trview
+                    return 1;
+                }
+                else if (key == "inferred_rooms")
+                {
+                    if (auto level = camera_sink->level().lock())
+                    {
+                        std::vector<std::weak_ptr<IRoom>> inferred_rooms;
+                        std::ranges::transform(camera_sink->inferred_rooms(),
+                            std::back_inserter(inferred_rooms),
+                            [&level](const auto& r) { return level->room(r); });
+                        return push_list(L, inferred_rooms, { create_room });
+                    }
+                }
+                else if (key == "number")
                 {
                     lua_pushinteger(L, camera_sink->number());
                     return 1;
                 }
+                else if (key == "persistent")
+                {
+                    lua_pushboolean(L, camera_sink->persistent());
+                    return 1;
+                }
+                else if (key == "position")
+                {
+                    create_vector3(L, camera_sink->position() * trlevel::Scale);
+                    return 1;
+                }
+                else if (key == "room")
+                {
+                    if (auto level = camera_sink->level().lock())
+                    {
+                        return create_room(L, level->room(camera_sink->room()).lock());
+                    }
+                }
+                else if (key == "strength")
+                {
+                    lua_pushinteger(L, camera_sink->strength());
+                    return 1;
+                }
+                else if (key == "type")
+                {
+                    lua_pushstring(L, to_string(camera_sink->type()).c_str());
+                    return 1;
+                }
+                else if (key == "triggered_by")
+                {
+                    return push_list(L, camera_sink->triggers(), { create_trigger });
+                }
+                else if (key == "visible")
+                {
+                    lua_pushboolean(L, camera_sink->visible());
+                    return 1;
+                }
                 return 0;
             }
 
             int camera_sink_newindex(lua_State* L)
             {
                 ICameraSink* camera_sink = *static_cast<ICameraSink**>(lua_touserdata(L, 1));
-                camera_sink;
+
+                const std::string key = lua_tostring(L, 2);
+                if (key == "type")
+                {
+                    if (auto level = camera_sink->level().lock())
+                    {
+                        const char* type_str = lua_tostring(L, -1);
+                        if (type_str)
+                        {
+                            std::string type = type_str;
+                            camera_sink->set_type(type == "Camera" ? ICameraSink::Type::Camera : ICameraSink::Type::Sink);
```

Comment:
Is this right? If the user types "camera", it's gonna get turned into a sink. Maybe throw an error if the input is not "Camera" or "Sink" instead.

Reply:
Added errors

Reply:
I have another PR to go through after this one that makes Lua build as C++ instead of C. At the moment if the error happened it would longjmp away and never return, so I don't think the destructors for the smart pointers would fire. Would end up with leaks.